default/scripting/common/enqueue.macros

```diff
@@ -1,3 +1,33 @@
+/* Allows the current building to be enqueued if the given prerequisite is built or enqueued.
+   Takes the prerequisite name as parameter; Usage:
+        [[LOCATION_ALLOW_ENQUEUE_IF_ENQUEUED(BLD_SHIPYARD_BASE)]]
+*/
+LOCATION_ALLOW_ENQUEUE_IF_ENQUEUED
+'''Or  [
+            Contains And [
+                Building name = "@1@"
+                OwnedBy empire = Source.Owner
+            ]
+            // Allows enqueue if this is not enqueued but prerequisite @1@ is
+            And [
+                Enqueued type = Building name = "@1@"
+                Not Enqueued type = Building name = CurrentContent
+            ]
+        ]'''
+/* Allows the current building to be build if the given prerequisite is built or enqueued.
+   Takes the prerequisite name as parameter; Usage:
+        [[LOCATION_ALLOW_BUILD_IF_ENQUEUED(BLD_SHIPYARD_ORBITAL_DRYDOCK)]]
+*/
+LOCATION_ALLOW_BUILD_IF_ENQUEUED
```

Comment:
I don't think we should really have/use this macro/requirement -- even though this may correspond to how some things had previously worked, that it was fully sufficient if the prereq were simply enqueued whether or not it were ever finished, I think that was just a bit sloppy, and we allowed it only because we hadn't yet figured out a better way to get the flexibility that we actually wanted.  IMO prereqs should have to be completed before the dependent structures can make progress (i.e., use the other new macro), or else they just shouldn't be prereqs.   (Perhaps that's what you asked about before and I just hadn't read closely enough.)

Reply:
I know that the way it worked was we tried to achieve what we actually achieve with the changes in this thread. But there may be a downside to using ALLOW_ENQUEUE_IF_ENQUEUED everywhere which we should check first: is the location condition correctly reevaluated when calculating future production?

Reply:
Something enqueued in reliance on its prereq being enqueued, with ALLOW_ENQUEUE_IF_ENQUEUED in its location condition, will be listed in the production queue as Never completing, until such point as the prereq is completed and then it will start being able to receive PP and will start showing a predicted completion time.  The prediction code is not going to be able to simulate the future completion of prereqs (at least, not without a significant restructuring/recoding of how building prereqs are specified and a very significant recoding of the production simulator).

I don't think that the ability to predict completion time before a prereq is completed is a sufficient reason to weaken the prereq condition down to only requiring that the prereq be enqueued.

Reply:
> The prediction code is not going to be able to simulate the future completion of prereqs

To elaborate, the location conditions test the gamestate, including objects in the universe and some other non-object-associated values. Simulating the production queue doesn't create any new objects in the universe, so a condition testing if an object exists can't change part way through the simulation.

Reply:
> I don't think that the ability to predict completion time before a prereq is completed is a sufficient reason to weaken the prereq condition down to only requiring that the prereq be enqueued.
I do not think this weakens the prereq in this case... we are talking shipyards and the real prerequisites are not the prerequisites of the yards but of the ability to build the ships. As the ships depend on all prerequisites you cant build a ship with one missing.

I think the only possible down sides of being able to build the prerequisites at the same time is that building in sequence takes longer (which i think is ok) and that the user has to understand which yards are necessary to build a ship (which i think is sufficiently understandable from the descriptions and from non enqueable condition info).
So i think these have to be balanced against the benefit of a more correct build queue prediction.

Reply:
> I think the only possible down sides of being able to build the prerequisites at the same time is that building in sequence takes longer (which i think is ok)

That issue might be more serious than you think. The minimum build time is as important a factor to correctly balance a game element as the build costs are. It makes sure that producing a certain item takes a certain minimum amount of turns, and that you can't circument this by throwing more PP at it.

Meaning, when it comes to shipyards, those minimum build times ensure that it takes you a certain time to build up a fully upgraded shipyard with all extensions, and that you cannot do this too quickly just by investing more PP.

The moment you allow the various shipyard buildings to be built in parallel, you effectively break that premise, making balancing the shipyard extension buildings more difficult.

Sure, we can just increase the minimum build time of the shipyard extenion buildings to account for the ability to build in parallel, but that will totally screw the case where you just want to add an extension to an already existing shipyard, as the build times would be much too long in case you do *not* build in parallel.

So I think this comes down to how important we think it is to maintain that it the time it takes you to build a fully upgraded shipyard must not be cut short, but stay at a certain level (to ensure proper game balance).

Reply:
@Vezzra Keeping the parallel build means not changing the balance. Dilvish was suggesting to go back to an older style, which would change balance. 

Back to the general discussion how we want the mechanic to work, I would rephrase what you said: it comes down to how important we think it is to be able to build advanced buildings faster if the prerequisites are already build

I also think that nobody complained about the balance concerning advanced shipyards since the parallel build was introduced, so i take that as a data point indicating that this is not serious.
So, i repeat myself: I think it is OK if we do not enforce sequential build for advanced shipyards.

Reply:
Sorry if I was a bit unclear: I merely wanted to point out that the issue of being able to build in parallel is a quite significant one in general, I wasn't specifically referring to the current change. The current state of things with the shipyards is a big mess anyway. Hence the lack of complaints about balance you mentioned - there was not much of well thought out balance to mess up to begin with.

For a proper fix/solution however what I said is relevant, IMO.